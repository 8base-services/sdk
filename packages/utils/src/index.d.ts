
export type FieldSchema = {
  id: string,
  name: string,
  displayName?: string,
  description?: string,
  fieldType: string,
  fieldTypeAttributes: Object,
  isSystem: boolean,
  isList: boolean,
  isMeta: boolean,
  isRequired: boolean,
  isUnique?: boolean,
  defaultValue: any,
  relation: {
    id: string,
    relationTableName: string,
    relationFieldName: string,
    refFieldName: string,
    refFieldDisplayName: string,
    refFieldIsList: boolean,
    refFieldIsRequired: boolean,
    refTable: {
      id: string,
      name: string,
    },
  },
};

export type TableSchema = {
  id: string,
  name: string,
  displayName?: string,
  isSystem?: boolean,
  fields: Array<FieldSchema>,
};

export type QueryGeneratorConfig = {
  deep?: number,
  withMeta?: boolean,
  relationItemsCount?: number,
  includeColumns?: null | string[],
};


type CreateQueryColumnsListConfig = {
  flatten?: boolean,
} & QueryGeneratorConfig


export type Schema = Array<TableSchema>;

export function createQueryColumnsList(
  tablesList: TableSchema[],
  tableName: string,
  config?: CreateQueryColumnsListConfig,
): Array<{ name: string, title: string, meta: Object }>;


export const FIELD_TYPE: {
  DATE: 'DATE',
  FILE: 'FILE',
  ID: 'ID',
  NUMBER: 'NUMBER',
  RELATION: 'RELATION',
  SWITCH: 'SWITCH',
  TEXT: 'TEXT',
  SMART: 'SMART',
  JSON: 'JSON',
};

export const SWITCH_FORMATS: {
  ON_OFF: 'ON_OFF',
  YES_NO: 'YES_NO',
  TRUE_FALSE: 'TRUE_FALSE',
  ACTIVE_INACTIVE: 'ACTIVE_INACTIVE',
  HIGH_LOW: 'HIGH_LOW',
  CUSTOM: 'CUSTOM',
};

export const SWITCH_VALUES: {
  ON_OFF: {
    true: 'On',
    false: 'Off',
  },
  YES_NO: {
    true: 'Yes',
    false: 'No',
  },
  TRUE_FALSE: {
    true: 'True',
    false: 'False',
  },
  ACTIVE_INACTIVE: {
    true: 'Active',
    false: 'Inactive',
  },
  HIGH_LOW: {
    true: 'High',
    false: 'Low',
  },
};

export const TEXT_FORMATS: {
  UNFORMATTED: 'UNFORMATTED',
  NAME: 'NAME',
  SSN: 'SSN',
  EIN: 'EIN',
  PHONE: 'PHONE',
  EMAIL: 'EMAIL',
};

export const NUMBER_FORMATS: {
  NUMBER: 'NUMBER',
  CURRENCY: 'CURRENCY',
  PERCENTAGE: 'PERCENTAGE',
  FRACTION: 'FRACTION',
  SCIENTIFIC: 'SCIENTIFIC',
};

export const FILE_FORMATS: {
  FILE: 'FILE',
  IMAGE: 'IMAGE',
};

export const SMART_FORMATS: {
  ADDRESS: 'ADDRESS',
  PHONE: 'PHONE',
};

export const DATE_FORMATS: {
  DATE: 'DATE',
  DATETIME: 'DATETIME',
};

export const MUTATION_TYPE: {
  CREATE: 'CREATE',
  UPDATE: 'UPDATE',
};

export const SYSTEM_TABLES: {
  USERS: 'Users',
  FILES: 'Files',
  SETTINGS: 'Settings',
  ROLES: 'Roles',
  INVITATIONS: 'Invitations',
  PERMISSIONS: 'Permissions',
  API_TOKENS: 'ApiTokens',
  ENVIRONMENT_VARIABLES: 'EnvironmentVariables',
};


export type MutationType = 'CREATE' | 'UPDATE';

export type FieldType = keyof typeof FIELD_TYPE;

export type Format = keyof typeof SWITCH_FORMATS
  | keyof typeof TEXT_FORMATS
  | keyof typeof NUMBER_FORMATS
  | keyof typeof FILE_FORMATS
  | keyof typeof SMART_FORMATS
  | keyof typeof DATE_FORMATS;

export const MUTATION_FILE_FIELDS: ['fileId', 'public', 'filename', 'meta', 'mods'];

export const tableSelectors: { [key: string]: (schema: TableSchema, arg?: any) => any };

export const tableFieldSelectors: { [key: string]: (field: FieldSchema, arg?: any) => any };

export const getTableSchemaByName: (schema: TableSchema[], name: string) => TableSchema;

export const getTableSchemaById: (schema: TableSchema[], id: string) => TableSchema;

export const throwIfMissingRequiredParameters: Function;

export const formatDataForMutation: Function;

export const createTableRowCreateTag: Function;
export const createTableRowCreateManyTag: Function;
export const createTableRowUpdateTag: Function;
export const createTableRowDeleteTag: Function;
export const createTableRowQueryTag: Function;
export const createTableFilterGraphqlTag: Function;

export type AuthState = {
  email?: string,
  userId?: string,
  workspaceId?: string,
  refreshToken?: string,
  token?: string,
}

export type AuthData = {
  isEmailVerified: boolean,
  idToken: string,
  email: string,
  idTokenPayload: Object,
  state: any,
};

export interface AuthClient {
  getAuthState(): Promise<AuthState>,
  setAuthState(state: AuthState): Promise<void>,
  purgeAuthState(): Promise<void>,
  checkIsAuthorized(): Promise<boolean>,
}

export interface Authorizable {
  authorize(options?: Object): Promise<AuthData>,
  renewToken(options?: Object): Promise<AuthData>,
  changePassword(): Promise<{ email: string }>,
  logout(options?: Object): Promise<void>,
}

export class SDKError extends Error {
  name: string;
  code: string;
  packageName: string;
  message: string;

  constructor(code: string, packageName: string, message: string);

  toString(): string;

  toJSON(): {};
}

export const ERROR_CODES: {
  UNSUPPORTED_FIELD_TYPE: 'UnsupportedFieldType',
  MISSING_PARAMETER: 'MissingParameter',
};

