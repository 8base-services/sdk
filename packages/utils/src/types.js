// @flow
import {
  FIELD_TYPE,
  SWITCH_FORMATS,
  TEXT_FORMATS,
  NUMBER_FORMATS,
  FILE_FORMATS,
  DATE_FORMATS,
} from './constants';

type MutationType = 'CREATE' | 'UPDATE';

type FieldType = $Values<typeof FIELD_TYPE>;

type Format = $Values<typeof SWITCH_FORMATS>
  | $Values<typeof TEXT_FORMATS>
  | $Values<typeof NUMBER_FORMATS>
  | $Values<typeof FILE_FORMATS>
  | $Values<typeof DATE_FORMATS>;

type FieldSchema = {
  id: string,
  name: string,
  displayName?: string,
  description: ?string,
  fieldType: FieldType,
  fieldTypeAttributes: Object,
  isSystem: boolean,
  isList: boolean,
  isMeta: boolean,
  isRequired: boolean,
  isUnique: ?boolean,
  defaultValue: any,
  relation: {
    id: string,
    relationTableName: string,
    relationFieldName: string,
    refFieldName: string,
    refFieldDisplayName: string,
    refFieldIsList: boolean,
    refFieldIsRequired: boolean,
    refTable: {
      id: string,
    },
  },
};

type TableSchema = {
  id: string,
  name: string,
  displayName?: string,
  isSystem: boolean,
  fields: Array<FieldSchema>,
};

type Schema = Array<TableSchema>;

export type {
  MutationType,
  FieldType,
  Format,
  FieldSchema,
  TableSchema,
  Schema,
};
