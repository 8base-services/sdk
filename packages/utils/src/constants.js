//@flow

const FIELD_TYPE = {
  DATE: 'DATE',
  FILE: 'FILE',
  ID: 'ID',
  NUMBER: 'NUMBER',
  RELATION: 'RELATION',
  SWITCH: 'SWITCH',
  TEXT: 'TEXT',
};

const FORMAT = {
  UNFORMATTED: 'UNFORMATTED',
  NAME: 'NAME',
  ADDRESS: 'ADDRESS',
  SSN: 'SSN',
  EIN: 'EIN',
  PHONE: 'PHONE',
  EMAIL: 'EMAIL',
  DATE: 'DATE',
  DATETIME: 'DATETIME',
};

const MUTATION_TYPE = {
  CREATE: 'CREATE',
  UPDATE: 'UPDATE',
};

export {
  FORMAT,
  FIELD_TYPE,
  MUTATION_TYPE,
};
