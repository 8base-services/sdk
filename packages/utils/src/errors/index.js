export { SDKError } from './SDKError';
export { ERROR_CODES } from './codes';
export { throwIfMissingRequiredParameters } from './helpers';

