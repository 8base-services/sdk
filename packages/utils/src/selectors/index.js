// @flow

import * as tableSelectors from './tableSelectors';
import * as tableFieldSelectors from './tableFieldSelectors';

export { getTableSchemaByName } from './getTableSchemaByName';
export { getTableSchemaById } from './getTableSchemaById';

export {
  tableSelectors,
  tableFieldSelectors,
};

