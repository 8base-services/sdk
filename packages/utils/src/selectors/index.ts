import * as tableSelectors from './tableSelectors';
import * as tablesListSelectors from './tablesListSelectors';
import * as tableFieldSelectors from './tableFieldSelectors';
 
export { tableSelectors, tableFieldSelectors, tablesListSelectors };
