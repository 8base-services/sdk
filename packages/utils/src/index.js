import * as tableSelectors from './selectors/tableSelectors';

export * from './constants';
export * from './formatters';
export * from './selectors';
export * from './verifiers';
export * from './queryTableGenerator';
export * from './functionParameters';

export { tableSelectors };
