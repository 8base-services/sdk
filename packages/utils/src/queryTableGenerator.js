// @flow
import * as R from 'ramda';
import { createSelector } from 'reselect';
import { SchemaNameGenerator } from '@8base/sdk';

import * as tableSelectors from './selectors/tableSelectors';
import * as tableFieldSelectors from './selectors/tableFieldSelectors';
import type { TableSchema } from './types';

export const TABLE_CONTENT_NAME = 'tableContent';

export type CheckedRule = {
  id: string,
  name: string,
  checked: boolean,
}

type QueryObjectConfig = {
  deep?: number,
  withMeta?: boolean,
  includeColumns?: string[],
}

type QueryStringConfig = {
  prevSpaceCount?: number,
  spaceCount?: number,
  initSpaceCount?: number,
} & QueryObjectConfig


const upperFirst = (str: string) => R.toUpper(R.head(str)) + R.tail(str);

const getTableByName = (tablesList: TableSchema[], tableName: string) =>
  tablesList.find(({ name }) => tableName === name);


const emptyRelation = {
  id: 'id',
  _description: '_description',
};

const emptyRelationList = {
  items: {
    id: 'id',
    _description: '_description',
  },
  count: 'count',
};

export const getQueryObject = (tablesList: TableSchema[], tableName: string, queryObjectConfig: QueryObjectConfig = {}, prevKey?: string = '') => {
  const { fields = [] } = getTableByName(tablesList, tableName) || {};
  const { deep = 3, withMeta = true, includeColumns } = queryObjectConfig;

  const queryObject = {};
  fields
    .filter((field) => {
      const isMeta = tableFieldSelectors.isMetaField(field);

      return withMeta
        ? true
        : !isMeta;
    })
    .forEach(field => {
      let fieldContent = field.name;
      const isRelation = tableFieldSelectors.isRelationField(field);
      const isFile = tableFieldSelectors.isFileField(field);
      const isSmart = tableFieldSelectors.isSmartField(field);
      const isList = tableFieldSelectors.isListField(field);
      const isSettingsRefTable = tableFieldSelectors.isSettingsRefTable(field);
      const refTableName = tableFieldSelectors.getRelationTableName(field);
      const refTable = getTableByName(tablesList, refTableName);
      const currentKeyString = prevKey ? `${prevKey}.${field.name}` : field.name;

      if (isSettingsRefTable) {
        fieldContent = {
          _description: '_description',
        };
      } else if (isRelation) {

        if (deep > 1) {
          if (!!refTableName && !!refTable) {
            const includeAllrelationFields = R.contains(currentKeyString, includeColumns);
            const relationIncludeColumns = includeAllrelationFields
              ? null
              : includeColumns;

            fieldContent = {
              id: 'id',
              ...getQueryObject(tablesList, refTableName, { deep: deep - 1, withMeta, includeColumns: relationIncludeColumns }, currentKeyString),
              _description: '_description',
            };
          } else {
            fieldContent = {
              id: 'id',
              _description: '_description',
            };
          }
        } else {
          fieldContent = null;
        }
      } else if (isFile) {
        fieldContent = {
          id: 'id',
          fileId: 'fileId',
          filename: 'filename',
          downloadUrl: 'downloadUrl',
          shareUrl: 'shareUrl',
          meta: 'meta',
        };
      } else if (isSmart) {
        fieldContent = field.fieldTypeAttributes.innerFields.reduce(
          (accum, { name }) => { accum[name] = name; return accum; },
          {},
        );
      }

      if (isList && (isRelation || isFile) && fieldContent !== null) {
        fieldContent = {
          items: fieldContent,
          count: 'count',
        };
      }

      const isNotEmptyrelation = isRelation && !R.equals(fieldContent, emptyRelation) && !R.equals(fieldContent, emptyRelationList);

      const needsInclude = !!includeColumns
        ? R.contains(currentKeyString, includeColumns)
        : true;

      console.log(currentKeyString, needsInclude, isNotEmptyrelation, includeColumns);

      if (fieldContent !== null && (needsInclude || isNotEmptyrelation)) {
        queryObject[field.name] = fieldContent;
      }
    });

  return queryObject;
};

const capitalizeFirstLetter = (str: string) => {
  return str.charAt(0).toUpperCase() + str.slice(1);
};

const transformQueryObjectToString = (queryObject: Object, spacesParams: *) => {
  const { prevSpaceCount = 4, spaceCount = 2, initSpaceCount = 2 } = spacesParams || {};

  const queryString = Object.keys(queryObject).reduce(
    (accum, queryElement) => {
      const spaces = R.repeat(' ', prevSpaceCount + spaceCount).join('');

      if (typeof queryObject[queryElement] === 'object') {
        const innerObjectString = transformQueryObjectToString(
          queryObject[queryElement],
          { prevSpaceCount, initSpaceCount, spaceCount: initSpaceCount + spaceCount },
        );

        accum += `\n${spaces}${queryElement} {${innerObjectString}\n${spaces}}`;
      } else {
        accum += `\n${spaces}${queryObject[queryElement]}`;
      }

      return accum;
    },
    '',
  );

  return queryString;
};


export const createQueryColumnsList = (tablesList: TableSchema[], tableName: string, queryObjectConfig: QueryObjectConfig = {}) => {
  const { fields = [] } = getTableByName(tablesList, tableName) || {};
  const { deep = 3, withMeta = false, includeColumns } = queryObjectConfig;

  const transformedList = fields
    .filter((field) => {
      const isMeta = tableFieldSelectors.isMetaField(field);

      return withMeta
        ? true
        : !isMeta;
    })
    .map((field) => {
      const fieldName = tableFieldSelectors.getFieldName(field);
      const fieldType = tableFieldSelectors.getFieldType(field);
      const isRelation = tableFieldSelectors.isRelationField(field);
      const isFile = tableFieldSelectors.isFileField(field);
      const isList = tableFieldSelectors.isListField(field);
      const refTableName = tableFieldSelectors.getRelationTableName(field);

      const title = capitalizeFirstLetter(fieldName);
      const refTable = getTableByName(tablesList, refTableName);

      const meta = {
        isList,
        fieldType,
      };

      if (isFile) {
        return [];
      } else if (isRelation && isList) {
        return [{
          name: fieldName, title, meta,
        }];
      } else if (isRelation && refTableName && refTable && deep > 1) {
        const innerKeys = createQueryColumnsList(tablesList, refTableName, { deep: deep - 1, withMeta, includeColumns });

        return innerKeys.map(({ name, ...rest }) => ({ ...rest, name: `${fieldName}.${name}` }));
      } else if (isRelation) {
        return [{
          name: fieldName, title, meta,
        }];
      }
      return [{
        name: fieldName, title, meta,
      }];
    });

  return R.flatten(transformedList)
    .filter(({ name }) => !!includeColumns
      ? R.contains(name, includeColumns)
      : true,
    );
};


export const createQueryString =
  (tablesList: TableSchema[], tableName: string, queryStringConfig?: QueryStringConfig = {}) => {
    const { prevSpaceCount, spaceCount, ...rest } = queryStringConfig;

    const gueryObject = getQueryObject(tablesList, tableName, {
      ...rest,
    });

    return transformQueryObjectToString(gueryObject, { prevSpaceCount, initSpaceCount: spaceCount, spaceCount });
  };

export const createQueryStringWithoutMetaFields =
  (tablesList: TableSchema[], tableName: string, queryStringConfig?: QueryStringConfig = {}) => {
    const { prevSpaceCount, spaceCount, ...rest } = queryStringConfig;

    const gueryObject = getQueryObject(tablesList, tableName, {
      withMeta: false,
      ...rest,
    });

    return transformQueryObjectToString(gueryObject, { prevSpaceCount, initSpaceCount: spaceCount, spaceCount });
  };

export const createTableFilterGraphqlTag = (tablesList: TableSchema[], tableName: string, config: * = {}) => `
  query DataViewerTable${upperFirst(tableName)}Content($filter: ${SchemaNameGenerator.getFilterInputTypeName(tableName)}, $orderBy: [${SchemaNameGenerator.getOrderByInputTypeName(tableName)}], $after: String, $before: String, $first: Int, $last: Int, $skip: Int) { 
    ${TABLE_CONTENT_NAME}: ${SchemaNameGenerator.getTableListFieldName(tableName)}(filter: $filter, orderBy: $orderBy, after: $after, before: $before, first: $first, last: $last, skip: $skip) {
      items {${createQueryString(tablesList, tableName, { ...config, prevSpaceCount: 6 })}
        _description
      }
      count
    }
  }`;

export const createTableRowCreateTag = (tablesList: TableSchema[], tableName: string) => {
  const table = getTableByName(tablesList, tableName);
  const hasNonMetaFields = tableSelectors.hasNonMetaFields(table);

  if (hasNonMetaFields) {
    return `
  mutation DataViewer${upperFirst(tableName)}RowCreate($data: ${SchemaNameGenerator.getCreateInputName(tableName)}!) {
    ${SchemaNameGenerator.getCreateItemFieldName(tableName)}(data: $data) {
      id${createQueryStringWithoutMetaFields(tablesList, tableName)}
    }
  }`;
  }

  return `
  mutation DataViewer${upperFirst(tableName)}RowCreate {
    ${SchemaNameGenerator.getCreateItemFieldName(tableName)} {
      id${createQueryStringWithoutMetaFields(tablesList, tableName)}
    }
  }`;
};

export const createTableRowUpdateTag = (tablesList: TableSchema[], tableName: string) => `
  mutation DataViewer${upperFirst(tableName)}RowUpdate($data: ${SchemaNameGenerator.getUpdateInputName(tableName)}!) {
    ${SchemaNameGenerator.getUpdateItemFieldName(tableName)}(data: $data) {
      id${createQueryStringWithoutMetaFields(tablesList, tableName)}
    }
  }`;

export const createTableRowQueryTag = (tablesList: TableSchema[], tableName: string) => `
  query DataViewer${upperFirst(tableName)}Row($id: ID!) {
    ${SchemaNameGenerator.getTableItemFieldName(tableName)}(id: $id) {${createQueryString(tablesList, tableName)}
    }
  }`;

export const createTableRowDeleteTag = (tablesList: TableSchema[], tableName: string) => `
  mutation DataViewer${upperFirst(tableName)}RowDelete($data: ${SchemaNameGenerator.getDeleteInputName(tableName)}!) {
    ${SchemaNameGenerator.getDeleteItemFieldName(tableName)}(data: $data) {
      success
    }
  }`;
