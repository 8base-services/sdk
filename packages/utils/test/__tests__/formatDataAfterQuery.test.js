//@flow

import { formatDataAfterQuery } from '../../src';
import { SCHEMA } from '../__fixtures__';

describe('As developer, I can format for create mutation,', () => {
  it('Compelex data.', () => {
    const data = {
      relationList: {
        items: [{
          scalar: 'Relation List Scalar Value',
          scalarList: [
            'Relation List Scalar List Value',
          ],
          nestedRelation: '5b32159b66a450c047285628',
        }],
      },
    };

    expect(formatDataAfterQuery('tableSchema', data, SCHEMA)).toMatchSnapshot();
  });
});
