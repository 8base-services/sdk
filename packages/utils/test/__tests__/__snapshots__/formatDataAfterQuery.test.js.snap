// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`As developer, I can format for create mutation, Compelex data. 1`] = `
Object {
  "address": Object {
    "city": "Kenia Urhahn",
    "state": "Scottie Swailes",
    "street1": "Pamelia Quall",
    "street2": "Lasonya Friedly",
    "zip": "Timothy Ingleton",
  },
  "meta": "meta",
  "relation": Object {
    "scalar": "Relation Scalar Value",
  },
  "relationList": Array [
    Object {
      "nestedRelation": "5b32159b66a450c047285628",
      "scalar": "Relation List Scalar Value",
      "scalarList": Array [
        "Relation List Scalar List Value",
      ],
    },
  ],
  "scalar": "Scalar Value",
  "scalarList": Array [
    "Scalar List Value",
  ],
}
`;
